ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -333,6 +428,118 @@ impl Module {
+        ExtraArgsIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Lookup a type by its index. Panics if an out of bounds index is supplied.
+    fn get_type(&self, tidx: TypeIndex) -> &Type {
+        &self.types[usize::from(tidx.0)]
+    }
+
+    /// Push a new type into the type table and return its index.
+    fn push_type(&mut self, ty: Type) -> TypeIndex {
+        let idx = self.types.len();
+        self.types.push(ty);
+        TypeIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Push a new constant into the constant table and return its index.
+    pub(crate) fn push_const(&mut self, constant: Constant) -> ConstIndex {
+        let idx = self.consts.len();
+        self.consts.push(constant);
+        ConstIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Get the index of a type, inserting it in the type table if necessary.
+    pub fn const_index(&mut self, c: &Constant) -> ConstIndex {
+        // FIXME: can we optimise this?
+        for (idx, tc) in self.consts.iter().enumerate() {
+            if tc == c {
+                // const table hit.
+                return ConstIndex(u16::try_from(idx).unwrap()); // FIXME: propagate error
+            }
+        }
+        // type table miss, we need to insert it.
+        self.push_const(c.clone())
+    }
+
+    /// Get the index of a type, inserting it in the type table if necessary.
+    pub fn type_index(&mut self, ty: &Type) -> TypeIndex {
+        // FIXME: can we optimise this?
+        for (idx, tty) in self.types.iter().enumerate() {
+            if ty == tty {
+                // type table hit.
+                return TypeIndex(u16::try_from(idx).unwrap()); // FIXME: propagate error
+            }
+        }
+        // type table miss, we need to insert it.
+        self.push_type(ty.clone())
+    }
+
+    /// Push a function into the function table and return its index.
+    pub(crate) fn push_func(&mut self, func: Function) -> FuncIndex {
+        let idx = self.funcs.len();
+        self.funcs.push(func);
+        FuncIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Get the index of a function, inserting it into the function table if necessary.
+    pub(crate) fn func_index(&mut self, func: &Function) -> FuncIndex {
+        // FIXME: can we optimise this?
+        for (idx, tfunc) in self.funcs.iter().enumerate() {
+            // It's enough to just compare the name.
+            if tfunc.name == func.name {
+                // type table hit.
+                return FuncIndex(u16::try_from(idx).unwrap()); // FIXME: propagate error
+            }
+        }
+        // type table miss, we need to insert it.
+        self.push_func(func.clone())
+    }
+
+    /// Look up a [Function] by index. Panics if an out of bounds index is supplied.
+    pub(crate) fn get_func(&self, idx: FuncIndex) -> &Function {
```

Comment:
Why do we need `func_index` and `get_func`? I'm a bit unsure why I'd ever call `func_index` for a given function more than once.

Reply:
At the moment `get_func()` is used only in a debug assertion.

`func_index()` is used only once while creating the initial JIT IR.

I don't know the answer to your question, but I expect the API to refine itself as we discover the use-cases required of it.

Reply:
> At the moment get_func() is used only in a debug assertion.

Can we guard it with `#[cfg(debug_assertions)]` or similar?

> func_index() is used only once while creating the initial JIT IR.

Right, so my question is: do we ever call this for a function F more than once? I kinda guess "no", but maybe I'm wrong.

Reply:
`func_index()` and `get_func()` were deleted by the "sharing with AOT" commit: c8333b87fab7833bd6584ed4234b0d27b4c545db